src/Shared/ServerInfrastructure/BufferExtensions.cs

```diff
@@ -199,32 +199,60 @@ private static void WriteNumericMultiWrite(ref this BufferWriter<PipeWriter> buf
             buffer.Write(new ReadOnlySpan<byte>(byteBuffer, position, length));
         }
 
+        internal static void WriteEncoded(ref this BufferWriter<PipeWriter> buffer, string data, Encoding encoding)
+        {
+            if (string.IsNullOrEmpty(data))
+            {
+                return;
+            }
+
+            var dest = buffer.Span;
+            var sourceLength = encoding.GetByteCount(data);
+            // Fast path, try encoding to the available memory directly
+            if (sourceLength <= dest.Length)
+            {
+                encoding.GetBytes(data, dest);
+                buffer.Advance(sourceLength);
+            }
+            else
+            {
+                WriteEncodedMultiWrite(ref buffer, data, sourceLength, encoding);
+            }
+        }
+
         [MethodImpl(MethodImplOptions.NoInlining)]
-        private static void WriteAsciiMultiWrite(ref this BufferWriter<PipeWriter> buffer, string data)
+        private static void WriteEncodedMultiWrite(ref this BufferWriter<PipeWriter> buffer, string data, int encodedLength, Encoding encoding)
         {
-            var dataLength = data.Length;
-            var offset = 0;
+            var source = data.AsSpan();
             var bytes = buffer.Span;
-            do
+            var totalBytesUsed = 0;
+            var encoder = encoding.GetEncoder();
+            var minBufferSize = encoding.GetMaxByteCount(1);
+            var completed = false;
+
+            // This may be a bug, but encoder.Convert returns completed = true for UTF7 too early.
```

Comment:
Is there an issue filed for UTF7 so they can investigate?

Reply:
https://github.com/dotnet/runtime/issues/414